### ISSUE TITLE
[DataStore] Failed to execute 'index' error when deleting a record with `@hasMany` relationships

### ISSUE THREAD

**Issue by chrisbonifacio** [aws-amplify/amplify-js#10864]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

React

### Amplify APIs

DataStore

### Environment information

<details>

```
# Put output below this line


```

</details>

### Describe the bug

Getting an error from DataStore when trying to delete a record (with associated @hasMany related objects)
Model:

```graphql
type CutsheetRequirement @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetRequirementId: ID! @primaryKey
  name: String!
  ...
  CutsheetOptions: [CutsheetOption] @hasMany(indexName: "byCutsheetRequirement", fields: ["cutsheetRequirementId"])
  CutsheetImpacts: [CutsheetImpact] @hasMany(indexName: "byCutsheetRequirement", fields: ["cutsheetRequirementId"])
  ...
}

type CutsheetOption @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetOptionId: ID! @primaryKey
  name: String!
  cutsheetRequirementID: ID! @index(name: "byCutsheetRequirement", queryField: "listCutsheetOptionsByCutsheetRequirementId", sortKeyFields: ["cutsheetOptionId"])
  sort: Int!
  CutsheetImpacts: [CutsheetImpact] @hasMany(indexName: "byCutsheetOption", fields: ["cutsheetOptionId"])
  CutsheetInstructions: [CutsheetInstruction] @hasMany(indexName: "byCutsheetOption", fields: ["cutsheetOptionId"])
  updatedBy: String!
  updatedSource: String!
}

type CutsheetInstruction @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetInstructionId: ID! @primaryKey
  name: String!
  cutsheetOptionID: ID! @index(name: "byCutsheetOption", queryField: "listCutsheetInstructionsByCutsheetOptionId", sortKeyFields: ["cutsheetInstructionId"])
  subContractorType: SubContractorType!
  heading: String
  instructions: String!
  sort: Int!
  updatedBy: String!
  updatedSource: String!
}

type CutsheetImpact @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetImpactId: ID! @primaryKey
  cutsheetRequirementID: ID @index(name: "byCutsheetRequirement", queryField: "listCutsheetImpactsByCutsheetRequirementID", sortKeyFields: ["cutsheetImpactId"])
  cutsheetOptionID: ID @index(name: "byCutsheetOption", queryField: "listCutsheetImpactsByCutsheetOptionID", sortKeyFields: ["cutsheetImpactId"])
  actionType: CutsheetImpactActionType!
  updatedBy: String!
  updatedSource: String!
}
```

Delete command:
```js
  try {
    const toDelete = await DataStore.query(CutsheetRequirement, { cutsheetRequirementId: id});
    if (toDelete) {
      await DataStore.delete(toDelete);
    }
  } catch (e) {
    console.log('error', e);
  }
```

Error:
```js

error DOMException: Failed to execute 'index' on 'IDBObjectStore': The specified index was not found.
    at Proxy.<anonymous> (http://localhost:3000/static/js/bundle.js:42473:22)
    at IndexedDBAdapter.<anonymous> (http://localhost:3000/static/js/bundle.js:35975:101)
    at step (http://localhost:3000/static/js/bundle.js:159387:17)
    at Object.next (http://localhost:3000/static/js/bundle.js:159336:14)
    at fulfilled (http://localhost:3000/static/js/bundle.js:159295:24)
```

I have traced the error into IndexedDBAdapter.ts:943 where it seems to have loaded the index names incorrectly above on 928.

### Expected behavior

I expect record to delete without throwing an error.

### Reproduction steps

1. Deploy schema with given model
2. Create some records for each model
3. Attempt to delete a record with `@hasMany` relationships but no children

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

**Comment by rajczi** [comment]
The exception is thrown here as the index name it's using to find the children is incorrect (or undefined)
![image](https://user-images.githubusercontent.com/221660/212170768-54f73692-ae63-4c3d-b473-58c5e5869a46.png)

**Comment by rajczi** [comment]
There is a diagnositic report up there if it helps:
```
crajczi@MCWYDJV9H6 eh-data % amplify diagnose --send-report

Learn more at https://docs.amplify.aws/cli/reference/diagnose/

✅ Report saved: /var/folders/wr/dl7wd40s7pdbx2r1bxxpj0_w0000gn/T/ehdata/report-1673553329887.zip

✔ Done

Project Identifier: 55a68b1e5e23dd3cbfa9dc4b85af1b2d
```

**Comment by rajczi** [comment]
@iartemiev I've tried the following:

1. removing the indexName and fields properties (`@hasMany` without specifications   to default the index name and property) but that broke the relationships (child array was always empty in datastore and appsync though directly querying the related object by parentId returns results).
2. putting back just the field names (still no children when opening the relationship).
3. putting back index names but guaranteeing they are always unique (now the children show and the missing index name bug returns for cascade delete).

We're at a catch 22, can't get children without specifying the index, can't delete a model if the index is specified (with our without children since the index name doesn't load).

**Comment by chrisbonifacio** [comment]
Hi @rajczi 👋 our oncall is currently working on a fix for this issue.
Thank you for providing more context and what you've tried so far.

May be related to #10866

**Comment by iartemiev** [comment]
Hey @rajczi, I believe there's two issues at play here: 
1. In your schema you're attempting to connect your child models using a composite key index on the child to a single field on the parent. 

Ex:
```graphql
type CutsheetRequirement @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetRequirementId: ID! @primaryKey
  name: String!
  ...
  CutsheetOptions: [CutsheetOption] @hasMany(
    indexName: "byCutsheetRequirement", 
    fields: ["cutsheetRequirementId"]
  ) # single field `cutsheetRequirementId`
  ...
}

type CutsheetOption @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetOptionId: ID! @primaryKey
  cutsheetRequirementID: ID! @index(
    name: "byCutsheetRequirement", 
    queryField: "listCutsheetOptionsByCutsheetRequirementId", 
    sortKeyFields: ["cutsheetOptionId"]
  ) # index is composed of two fields: `cutsheetRequirementID` and `cutsheetOptionId`
}
```
When deleting `CutsheetRequirement`, DataStore will attempt to retrieve its children by that index, but since the parent doesn't have a field that maps to `cutsheetOptionId`, it can't resolve them, causing the error you're seeing. 

Unless you have a specific need to configure that index with a composite key, I suggest getting rid of the `sortKeyFields` attributes for each `@index` that you're using for connecting your models.

For the model I mentioned above, you should end up with:

```graphql
type CutsheetOption @model @auth(rules: [{ allow: private, provider: userPools }, { allow: public, provider: apiKey }]) {
  cutsheetOptionId: ID! @primaryKey
  cutsheetRequirementID: ID! @index(
    name: "byCutsheetRequirement", 
    queryField: "listCutsheetOptionsByCutsheetRequirementId", 
  ) 
}
```

2. There's currently a bug in the library when performing cascading deletes on nested `@hasMany` relations (the issue Chris linked https://github.com/aws-amplify/amplify-js/issues/10866). 
Once you modify your schema the way I suggested above, you should see that same error (Error: Namespace Resolver for 'Object' not found! This is probably a bug in '@amplify-js/datastore'). 
This bug is fixed in [this PR](https://github.com/aws-amplify/amplify-js/pull/10880) and will be released with the next version of AmplifyJS

**Comment by iartemiev** [comment]
@rajczi - a correction for what I said above 1. is not currently supported in DataStore, but we will treat it as a bug and will work on adding support in the near future.

However, if the schema change I proposed is acceptable for your use case, you can use it as a workaround in the interim. We just published a fix for the bug I mentioned in 2. with `aws-amplify@5.0.11` that will unblock you if you remove the `sortKeyFields` attribute and perform an `amplify push` after.

**Comment by rajczi** [comment]
I have already updated the schema as suggested and pushed the code.  I also confirmed we got the object missing exception as expected.  I will pull down 5.0.11 today and verify.That sortKeyFields entry was based on an example we pulled from the documentation and applied the pattern everywhere.  Sent from my iPhoneOn Jan 19, 2023, at 4:05 PM, Ivan Artemiev ***@***.***> wrote:﻿
@rajczi - a correction for what I said above 1. is not currently supported in DataStore, but we will treat it as a bug and will work on adding support in the near future.
However, if the schema change I proposed is acceptable for your use case, you can use it as a workaround in the interim. We just published a fix for the bug I mentioned in 2. with ***@***.*** that will unblock you if you remove the sortKeyFields attribute and perform an amplify push after.

**Comment by apolat2000** [comment]
I am still having the exact problem with `"@aws-amplify/datastore": "^4.0.16"` and `amplify": "^5.0.16"`. The model that I cannot delete with DataStore looks like:

```
type Entity @model @auth(rules: [{allow: owner, ownerField: "organization_id", identityClaim: "custom:organization_id"}]) {
  name: I18nString!
  description: I18nString!
  parentEntityID: ID
  level: Level! @hasOne
  location: Location
  customData: [AppliedCustomData]!
  appliedInterventions: [AppliedIntervention!]! @hasMany
  schemeVersion: Int @default(value: "0")
}
```

I can delete other models.

**Comment by iartemiev** [comment]
@apolat2000 could you share your `AppliedCustomData` and `AppliedIntervention` model definitions as well? I'll attempt to repro. I wouldn't expect this error to surface with an implicit `@hasMany` (i.e. without referencing an index).

**Comment by iartemiev** [comment]
Hey all, we merged a fix for this bug. It's currently available in `aws-amplify@5.0.24-unstableV5.3` and will be in the next release that goes out next week.

**Comment by iartemiev** [comment]
The fix was just released as part of `aws-amplify@5.0.24`
